Use Path for references instead of File

jean-philippe-martin opened the pull request:
Update multiple tools to use a Path instead of File.

Updated: FeatureWalker, ReadWalker, IntervalWalker, MultiVariantWalker, AnnotateIntervals, BaseRecalibrator, SelectVariants, SplitNCigarReads.

Some tools will need a change to ReferenceSequenceFileWalker in htsjdk, so I couldn't change those. Same for the writers.

Part of addressing issue #3709

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ReferenceInputArgumentCollection.java:
```diff
@@ -14,6 +14,8 @@
     /**
      * Get the reference file specified at the command line, creating the File object first if necessary.
+     *
+     * Use getReferencePath() instead.
```
Why not just remove it and use wherever it appears the `getReferencePath().toFile()`?

jean-philippe-martin replied:
That's not a bad idea, but leaving it in makes it super-easy to find the places I still have to update (using the IDE's "find calls"). The latter syntax would require a full-text search, which is slower and less convenient. Also, it's public so removing it might break some code that depends on GATK. What I was thinking instead was to mark it as @deprecated once we're done cleaning house.

lbergelson replied:
That's a good point.  We should definitely fix those usages as well, but we could do it in another PR if you don't want to fit it into this one.  It looks like it needs some changes to the spark tools / one of the CNV tools.

magicDGS replied:
Dependent code on GATK should be aware that using a beta version might change the API - I think that I am one of the affected by changes in API and I am find with them, because I accepted it from the beginning...

magicDGS replied:
And I agree that it can wait, but before the release to do not have support for deprecated things...

jean-philippe-martin replied:
Good points. I went ahead and deleted getReferenceFile. The Spark usages look a bit wonky now, someone may want to revisit in a later PR and perhaps simplify the logic there.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/ReferenceFileSource.java:
```diff
@@ -35,8 +48,7 @@
     public ReferenceFileSource(final File fastaFile) {
-        // Will throw a UserException if the .fai and/or .dict are missing
-        reference = CachingIndexedFastaSequenceFile.checkAndCreate(Utils.nonNull(fastaFile));
+        this(fastaFile.toPath());
```
This will throw an NPE if the `fastaFile`is null. Better: `this(Utils.nonNull(fastaFile).toPath())`

jean-philippe-martin replied:
Good call, done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/exome/pulldown/BayesianHetPulldownCalculator.java:
```diff
@@ -9,6 +9,7 @@
 import htsjdk.samtools.reference.ReferenceSequenceFileWalker;
 import htsjdk.samtools.util.IntervalList;
 import htsjdk.samtools.util.SamLocusIterator;
+import java.nio.file.Path;
```
Unused import.

jean-philippe-martin replied:
Removed.

jean-philippe-martin replied:
It's back now, but also it's used. I was able to switch `BayesianHetPulldownCalculator` to `Path`.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/ReferenceFileSource.java:
```diff
@@ -35,8 +48,7 @@
      * @param fastaFile reference fasta file
      */
     public ReferenceFileSource(final File fastaFile) {
```
I think we can just get rid of this method completely and switch to path in the tests that use it.

jean-philippe-martin replied:
Agreed, and done.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFile.java:
```diff
@@ -134,53 +138,53 @@ public CachingIndexedFastaSequenceFile(final File fasta) throws FileNotFoundExce
         }
 
-        final File indexFile = new File(fastaFile.getAbsolutePath() + ".fai");
+        final Path indexPath = Paths.get(URI.create(fastaPath.toUri().toString() + ".fai"));
```
good call!

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFile.java:
```diff
@@ -134,53 +138,53 @@ public CachingIndexedFastaSequenceFile(final File fasta) throws FileNotFoundExce
+        final Path indexPath = Paths.get(URI.create(fastaPath.toUri().toString() + ".fai"));
 
         // determine the name for the dict file
         // TODO: use the htsjdk method implemented in https://github.com/samtools/htsjdk/pull/774
```
Thanks! Done.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFile.java:
```diff
@@ -134,53 +138,53 @@ public CachingIndexedFastaSequenceFile(final File fasta) throws FileNotFoundExce
         // for creating these files.
-        if (!indexFile.exists()) {
-            throw new UserException.MissingReferenceFaiFile(indexFile, fastaFile);
+        if (!Files.exists(indexPath)) {
```
Agreed.

jean-philippe-martin commented:
@lbergelson I added a jimfs test to ReferenceFileSourceUnitTest. While I was there I also discovered there's not just `enging.ReferenceFileSource` but also `engine.datasources.ReferenceFileSource`. I updated the latter a little bit so it would work with `Path`s.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/ReferenceFileSourceUnitTest.java:
```diff
@@ -12,7 +16,61 @@
     @Test(expectedExceptions = UserException.MissingReference.class)
     public void testMissingReferenceFile() throws IOException {
-        new ReferenceFileSource(GATKBaseTest.getSafeNonExistentFile("NonExistentReference.fasta").getAbsolutePath());
+        new org.broadinstitute.hellbender.engine.datasources.ReferenceFileSource(
```
Wow, I didn't realize we had no direct tests for this before...

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/ReferenceFileSourceUnitTest.java:
```diff
@@ -12,7 +16,61 @@
+            final Path dictPath = jimfs.getPath("reference.dict");
+            Files.createFile(dictPath);
+
+            new org.broadinstitute.hellbender.engine.ReferenceFileSource(refPath);
```
Could you copy a real fasta into jimfs and then assert that we can actually get correct data out of it instead of just that it doesn't crash?

jean-philippe-martin replied:
Is there a reference test that reads from disk that I can use as a model?

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/ReferenceFileSourceUnitTest.java:
```diff
@@ -12,7 +16,61 @@
+            final Path dictPath = jimfs.getPath("reference.dict");
+            Files.createFile(dictPath);
+
+            new org.broadinstitute.hellbender.engine.datasources.ReferenceFileSource(refPath);
```
This is so confusing... We should fix this naming collision..  I opened a ticket for it #3926.

I think this test should probably be moved to a new test class ReferenceFileSourceUnitTest in the same package as that.  If it could assert that we can really read data as well that would be great.

jean-philippe-martin replied:
Agreed; I split the class in two accordingly. Remains to actually read the data.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/datasources/ReferenceFileSource.java:
```diff
@@ -16,33 +19,40 @@
     }
 
-    @Override
+        @Override
```
an extra tab got added here somehow

jean-philippe-martin replied:
Got it, fixed.

lbergelson commented:
@jean-philippe-martin Thanks, and thanks for catching that other class, it's bizarre that we have 2... I have a few more minor comments, and then 👍

lbergelson commented:
@jean-philippe-martin It looks like `ReferenceDataSourceUnitTest` is what's covering the `engine.ReferenceFileSource` functionality.  It should be easy to adapt one of those tests to run in jmfs.

jean-philippe-martin commented:
Ah yes indeed, I see. Would you like me to move these tests to ReferenceFileSourceUnitTest? That seems like a better home for them.

jean-philippe-martin commented:
Aha, one interesting wrinkle that came when trying to remove `getReferenceFile()` entirely in favor of `getReferencePath()` : It is valid to have references of the form `gg://foobar` (referencing the Google Genomics Reference API). You can apparently put that string in a `File` just fine, but a `Path` will try to find the matching provider, and will fail since there is no NIO provider for "gg" (it doesn't behave like a filesystem).

The work around is to use getReferenceFileName(). The risk is that some places that allowed a relative path (relying on the call to getAbsolutePath) may now require an absolute path.

jean-philippe-martin commented:
OK, Travis passes. I think I've addressed all the comments, so the ball is in your camp!

lbergelson commented:
@jean-philippe-martin I think this seems good now.  There were some merge conflicts so I resolved those.  Hopefully tests will pass on the new version and we can merge.

lbergelson commented:
Aaah, there are more conflicts now

droazen commented:
I'll rebase this branch, since there are conflicts once again.
